components/authz-service/engine/opa/opa_internal_test.go

```diff
@@ -549,3 +429,866 @@ func randomPolicies(i int) []map[string]interface{} {
+			"policies": policyMap,
+			"roles":    roleMap,
+		})
+
+		err = s.makeAuthorizedProjectPreparedQuery(ctx)
+		require.NoError(b, err, "prepared authorized project query")
+
+		b.Run(fmt.Sprintf("store with %d projects %d policies, and %d roles", projCount, len(policyMap), roleCount), func(b *testing.B) {
+			var resp []string
+			var err error
+			for n := 0; n < b.N; n++ {
+				resp, err = s.V2FilterAuthorizedProjects(ctx, []string{member})
+				if err != nil {
+					b.Error(err)
+				}
+			}
+			projectsResponse = resp
+		})
+	}
+}
+
+func BenchmarkV2ProjectsAuthorizedWithIncreasingSubjects(b *testing.B) {
+	ctx := context.Background()
+
+	l, err := logger.NewLogger("text", "debug")
+	require.NoError(b, err, "init logger")
+	s, err := New(ctx, l)
+	require.NoError(b, err, "init state")
+
+	// keep these values constant as we increase the number of subjects
+	policyCount := 20
+	roleCount := 10
+
+	policies, roles := v2BaselineAndRandomPoliciesAndRoles(policyCount, roleCount)
+	s.v2p1Store = inmem.NewFromObject(map[string]interface{}{
+		"policies": policies,
+		"roles":    roles,
+	})
+
+	err = s.makeAuthorizedProjectPreparedQuery(ctx)
+	require.NoError(b, err, "prepared authorized project query")
+
+	subjectCount := []int{0, 1, 10, 30, 50, 100, 150, 300, 500, 1000, 10000}
+	for _, count := range subjectCount {
+		b.Run(fmt.Sprintf("input with %d subjects", count), func(b *testing.B) {
+			var resp []string
+			var err error
+			for n := 0; n < b.N; n++ {
+				resp, err = s.V2ProjectsAuthorized(ctx, []string{"user:local:test"}, "iam:projects:delete", "iam:projects", allProjects)
+				if err != nil {
+					b.Error(err)
+				}
+			}
+			projectsResponse = resp
+		})
+	}
+}
+
+func BenchmarkV2FilterAuthorizedProjectsWithIncreasingSubjects(b *testing.B) {
+	ctx := context.Background()
+
+	l, err := logger.NewLogger("text", "debug")
+	require.NoError(b, err, "init logger")
+	s, err := New(ctx, l)
+	require.NoError(b, err, "init state")
+
+	// keep these values constant as we increase the number of subjects
+	policyCount := 20
+	roleCount := 10
+
+	policies, roles := v2BaselineAndRandomPoliciesAndRoles(policyCount, roleCount)
+	s.v2p1Store = inmem.NewFromObject(map[string]interface{}{
+		"policies": policies,
+		"roles":    roles,
+	})
+
+	err = s.makeAuthorizedProjectPreparedQuery(ctx)
+	require.NoError(b, err, "prepared authorized project query")
+
+	subjectCount := []int{0, 1, 10, 30, 50, 100, 150, 300, 500, 1000, 10000}
+	for _, count := range subjectCount {
+		b.Run(fmt.Sprintf("input with %d subjects", count), func(b *testing.B) {
+			var resp []string
+			var err error
+			for n := 0; n < b.N; n++ {
+				resp, err = s.V2FilterAuthorizedProjects(ctx, append([]string{"user:local:test"}, randomTeams(count)...))
+				if err != nil {
+					b.Error(err)
+				}
+			}
+			projectsResponse = resp
+		})
+	}
+}
+
+// Q: What happens if the subject appears more often as a member of different policies?
```

Comment:
This was a bit of a tangent but it's a good example of a question we can answer with benchmarking.